### PR TITLE
Tour of async

### LINKED ISSUE #4379: Add Notes about Async/Await All the Way Up
## Issues with Existing Topics

On the [Asynchronous Programming page](https://docs.microsoft.com/en-us/dotnet/csharp/async), there is no content about using async/await all the way through the call stack.  IMHO, this is an important concept and not everybody might understand this.

The topic is discussed in [this Stack Overflow post](https://stackoverflow.com/a/29809054/185123).

## Requests for new Topics

1. Maybe under [important info and advice](https://docs.microsoft.com/en-us/dotnet/csharp/async#important-info-and-advice)?

2. Write an abstract

   - Async methods need to be used through the entire call stack otherwise they will be blocking.

### LINKED ISSUE #4973: New topics: A guided tour of the Task Asynchronous Protocol (TAP)
Async programming is a complicated topic. While `async`, `await` and the `Task` based protocols do a lot to make it more approachable, as developers get beyond the basic scenarios. MicrosoftDocs/azure-docs#5963 has a lengthy discussion that surfaces many of the questions developers have about async and await:

1. The distinctions between a synchronous method that returns an awaitable object (like `Task`) and an async method is subtle. Developers can be confused about when the `async` modifier is needed. Developers have concerns about possible performance implications of the async state machine.
1. Asynchronous programming and multi-threaded programming are related concepts. The nuances cause confusion for developers. Task based asynchrony vs. TPL (Task Parallel Library) vs. Threads. Developers need more clarity on the distinction between these tools and the scenarios they target.
1. More on async vs. parallel: Developers need more guidance on how async can increase throughput and efficient use of the threadpool resources.
1. `SynchronizationContext` is not needed for basic scenarios, but surfaces in many intermediate and advanced uses. Developers need more tutorials that show how this is used.
1. While on the subject of `SynchronizationContext`, developers have questions on context affinity and when `ConfigureAwait` is needed.

We have a topic on [async and await](https://docs.microsoft.com/dotnet/csharp/programming-guide/concepts/async/) and another section on [async concepts](https://docs.microsoft.com/dotnet/csharp/programming-guide/concepts/async/). Some of the above topics are covered there, but may not be surfaced as well as needed.

I've added the discussion label to this issue. What else regarding async isn't covered yet that should be?  As you comment, note that there are other issues [related to async docs](https://github.com/dotnet/docs/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+async) already created.

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #4973
Fixes #4379 

This replaces the existing introduction to the async / await programming model with a new scenario based on a well-reviewed talk given at NDC

The previous introduction moves to a new article describing the asynchronous programming model.

### CODE REVIEW

File: docs/csharp/programming-guide/concepts/async/task-asynchronous-programming-model.md [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+  
+5.  `DoIndependentWork` is a synchronous method that does its work and returns to its caller.  
+  
+6.  `AccessTheWebAsync` has run out of work that it can do without a result from `getStringTask`. `AccessTheWebAsync` next wants to calculate and return the length of the downloaded string, but the method can't calculate that value until the method has the string.  
+  
+     Therefore, `AccessTheWebAsync` uses an await operator to suspend its progress and to yield control to the method that called `AccessTheWebAsync`. `AccessTheWebAsync` returns a `Task<int>` to the caller. The task represents a promise to produce an integer result that's the length of the downloaded string.  
+  
+    > [!NOTE]
+    >  If `GetStringAsync` (and therefore `getStringTask`) is complete before `AccessTheWebAsync` awaits it, control remains in `AccessTheWebAsync`. The expense of suspending and then returning to `AccessTheWebAsync` would be wasted if the called asynchronous process (`getStringTask`) has already completed and `AccessTheWebSync` doesn't have to wait for the final result.
```
Comment (rpetrusha) [inline docs/csharp/programming-guide/concepts/async/task-asynchronous-programming-model.md]:
```suggestion
    >  If `GetStringAsync` (and therefore `getStringTask`) completes before `AccessTheWebAsync` awaits it, control remains in `AccessTheWebAsync`. The expense of suspending and then returning to `AccessTheWebAsync` would be wasted if the called asynchronous process (`getStringTask`) has already completed and `AccessTheWebSync` doesn't have to wait for the final result.  
```

File: docs/csharp/programming-guide/concepts/async/task-asynchronous-programming-model.md [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+ `async` and `await` are contextual keywords. For more information and examples, see the following topics:  
+  
+-   [async](../../../../csharp/language-reference/keywords/async.md)  
+  
+-   [await](../../../../csharp/language-reference/keywords/await.md)  
+  
+## <a name="BKMK_ReturnTypesandParameters"></a> Return types and parameters  
+An async method typically returns a <xref:System.Threading.Tasks.Task> or a <xref:System.Threading.Tasks.Task%601>. Inside an async method, an `await` operator is applied to a task that's returned from a call to another async method.  
+  
+You specify <xref:System.Threading.Tasks.Task%601> as the return type if the method contains a [return](../../../../csharp/language-reference/keywords/return.md) statement that specifies an operand of type `TResult`.
```
Comment (rpetrusha) [inline docs/csharp/programming-guide/concepts/async/task-asynchronous-programming-model.md]:
```suggestion
You specify <xref:System.Threading.Tasks.Task%601> as the return type if the method contains a [`return`](../../../../csharp/language-reference/keywords/return.md) statement that specifies an operand of type `TResult`. 
```

File: docs/csharp/programming-guide/concepts/async/task-asynchronous-programming-model.md [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+  
+-   <xref:Windows.Foundation.IAsyncAction>, which corresponds to <xref:System.Threading.Tasks.Task>  
+  
+-   <xref:Windows.Foundation.IAsyncActionWithProgress%601>  
+  
+-   <xref:Windows.Foundation.IAsyncOperationWithProgress%602>  
+   
+  
+## <a name="BKMK_NamingConvention"></a> Naming convention  
+By convention, methods that return commonly awaitable types (e.g. `Task`, `Task<T>`, `ValueTask`, `ValueTask<T>`) should have names that end with "Async". Methods that start an asynchronous operation but do not return an awaitable type should not have names that end with "Async", but may start with "Begin", "Start" or some other verb to suggest this method does not return or throw the result of the operation.
```
Comment (rpetrusha) [inline docs/csharp/programming-guide/concepts/async/task-asynchronous-programming-model.md]:
```suggestion
By convention, methods that return commonly awaitable types (e.g. `Task`, `Task<T>`, `ValueTask`, `ValueTask<T>`) should have names that end with "Async". Methods that start an asynchronous operation but do not return an awaitable type should not have names that end with "Async", but may start with "Begin", "Start", or some other verb to suggest this method does not return or throw the result of the operation.
```

File: docs/csharp/programming-guide/concepts/async/index.md [hunk 4]
```diff
@@ -1,337 +1,141 @@
...
+
+Egg eggs = await eggTask;
+Console.WriteLine("eggs are ready");
+Bacon bacon = await baconTask;
+Console.WriteLine("bacon is ready");
+
+Console.WriteLine("Breakfast is ready!");
+```
+
+The preceding code works better. You start all the asynchronous tasks at once. You await each task only when you need the results. The preceding code may be similar to code in a web application that makes requests of different microservices, then combines the results into a single page. You'll make all the requests immediately. Then, `await` all those tasks, and compose the web page.
```
Comment (rpetrusha) [inline docs/csharp/programming-guide/concepts/async/index.md]:
```suggestion
The preceding code works better. You start all the asynchronous tasks at once. You await each task only when you need the results. The preceding code may be similar to code in a web application that makes requests of different microservices, then combines the results into a single page. You'll make all the requests immediately, then `await` all those tasks and compose the web page.
```

File: docs/csharp/programming-guide/concepts/async/index.md [hunk 5]
```diff
@@ -1,337 +1,141 @@
...
+
+## Composition with tasks
+
+ You have everything ready for breakfast at the same time. Except the toast. Making the toast is the composition of an asynchronous operation (toasting the bread), and synchronous operations (adding the butter and the jam). Updating this code illustrates an important concept:
```
Comment (rpetrusha) [inline docs/csharp/programming-guide/concepts/async/index.md]:
```suggestion
 You have everything ready for breakfast at the same time except the toast. Making the toast is the composition of an asynchronous operation (toasting the bread), and synchronous operations (adding the butter and the jam). Updating this code illustrates an important concept:
```

File: docs/csharp/programming-guide/concepts/async/index.md [hunk 6]
```diff
@@ -1,337 +1,141 @@
...
+
+> [!IMPORTANT]
+> The composition of an asynchronous operation followed by synchronous work is an asynchronous operation. Stated another way, if any portion of an operation is asynchronous, the entire operation is asynchronous.
+
+The preceding code showed you that you can use <xref:System.Threading.Tasks.Task> or <xref:System.Threading.Tasks.Task%601> objects to hold running tasks. You `await` each task before using its result. The next step is to create methods that represent the combination of other work. Before serving breakfast, you want to await the task that represents toasting the bread then adding butter and jam. You can represent that work with the following code:
```
Comment (rpetrusha) [inline docs/csharp/programming-guide/concepts/async/index.md]:
```suggestion
The preceding code showed you that you can use <xref:System.Threading.Tasks.Task> or <xref:System.Threading.Tasks.Task%601> objects to hold running tasks. You `await` each task before using its result. The next step is to create methods that represent the combination of other work. Before serving breakfast, you want to await the task that represents toasting the bread before adding butter and jam. You can represent that work with the following code:
```

File: docs/csharp/programming-guide/concepts/async/index.md [hunk 7]
```diff
@@ -1,337 +1,141 @@
...
+
+The preceding method has the `async` modifier in its signature. That signals to the compiler that this method contains an `await` statement; it contains asynchronous operations. This method represents the task that toasts the bread, then adds butter and jam. This method returns a <xref:System.Threading.Tasks.Task%601> that represents the composition of those three operations. The main block of code now becomes:
+
+[!code-csharp[StartConcurrentTasks](~/samples/snippets/csharp/tour-of-async/AsyncBreakfast-V3/Program.cs#Main)]
+
+The previous change illustrated an important technique for working with asynchronous code. You compose tasks by separating the operations into a new method that returns a task. You can choose when to await that task. You can start other tasks concurrently.
+
+## Await tasks efficiently
+
+The series of `await` statements at the end of the preceding code can be improved by using methods of the `Task` class. One of those APIs is <xref:System.Threading.Tasks.Task.WhenAll%2A>, which returns a <xref:System.Threading.Tasks.Task> that completes when all the tasks in its argument list have completed as shown in the following code:
```
Comment (rpetrusha) [inline docs/csharp/programming-guide/concepts/async/index.md]:
```suggestion
The series of `await` statements at the end of the preceding code can be improved by using methods of the `Task` class. One of those APIs is <xref:System.Threading.Tasks.Task.WhenAll%2A>, which returns a <xref:System.Threading.Tasks.Task> that completes when all the tasks in its argument list have completed, as shown in the following code:
```

File: docs/csharp/programming-guide/concepts/async/index.md [hunk 8]
```diff
@@ -1,337 +1,141 @@
...
+
+```csharp
+await Task.WhenAll(eggTask, baconTask, toastTask);
+Console.WriteLine("eggs are ready");
+Console.WriteLine("bacon is ready");
+Console.WriteLine("toast is ready");
+Console.WriteLine("Breakfast is ready!");
+```
+
+Another option is to use <xref:System.Threading.Tasks.Task.WhenAny%2A>, which returns a `Task<Task>` that completes when any of its arguments completes. You can await the returned task, knowing that it has already finished. The following code shows how you could use <xref:System.Threading.Tasks.Task.WhenAny%2A> to await the first task to finish, and then process its result. After processing the result from the completed task, you remove that completed task from the list of tasks passed to `WhenAny`.
```
Comment (rpetrusha) [inline docs/csharp/programming-guide/concepts/async/index.md]:
```suggestion
Another option is to use <xref:System.Threading.Tasks.Task.WhenAny%2A>, which returns a `Task<Task>` that completes when any of its arguments completes. You can await the returned task, knowing that it has already finished. The following code shows how you could use <xref:System.Threading.Tasks.Task.WhenAny%2A> to await the first task to finish and then process its result. After processing the result from the completed task, you remove that completed task from the list of tasks passed to `WhenAny`.
```

**Review by rpetrusha — APPROVED** [review]
This looks really good, @BillWagner. I've made some minor comments and suggestions, but you can merge when you're ready.

File: docs/csharp/programming-guide/concepts/async/task-asynchronous-programming-model.md [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+  
+ In async methods, you use the provided keywords and types to indicate what you want to do, and the compiler does the rest, including keeping track of what must happen when control returns to an await point in a suspended method. Some routine processes, such as loops and exception handling, can be difficult to handle in traditional asynchronous code. In an async method, you write these elements much as you would in a synchronous solution, and the problem is solved.  
+  
+ For more information about asynchrony in previous versions of the .NET Framework, see [TPL and Traditional .NET Framework Asynchronous Programming](../../../../standard/parallel-programming/tpl-and-traditional-async-programming.md).  
+  
+## <a name="BKMK_WhatHappensUnderstandinganAsyncMethod"></a> What happens in an async method  
+ The most important thing to understand in asynchronous programming is how the control flow moves from method to method. The following diagram leads you through the process.  
+  
+ ![Trace an async program](./media/task-asynchronous-programming-model/navigation-trace-async-program.png "NavigationTrace")
```
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/programming-guide/concepts/async/task-asynchronous-programming-model.md]: I can't take credit. this content already existed. I just renamed it.

**Pull request merged**